File: code/web/src/backend/kanjivg.ts
import { Option, bool, u8, Vec } from "./rust_compat";

export interface KanjiVG {
    kanji: Kanji[];
}

export interface Kanji {
    kanji: string,
    original: Option<string>,
    phon: Option<string>,
    position: Option<Position>,
    partial: bool,
    number: Option<u8>,
    part: Option<u8>,
    radical_form: bool,
    radical: Option<Radical>,
    strokes: Vec<Stroke>,
    parts: Vec<Kanji>,
    trad_form: bool,
    variant: bool,
}

export interface Stroke {
    path: string,
    typ: string,
}

export type Position = string;
export type Radical = string;

export function allPaths(kanji: Kanji): Stroke[] {
    return [
        ...(kanji.strokes ?? []),
        ...(kanji.parts?.flatMap(p => allPaths(p)) ?? []),
    ];
}

File: code/database/src/kanjivg.rs
use std::str::FromStr;

use serde::{Serialize, Deserialize};

#[derive(Default, Debug, Clone, Serialize, Deserialize)]
pub struct KanjiVG {
    pub kanji: Vec<Kanji>,
}

#[derive(Default, Debug, Clone, Serialize, Deserialize)]
pub struct Stroke {
    pub path: String,
    pub typ: String,
}

#[derive(Default, Debug, Clone, Serialize, Deserialize)]
pub struct Kanji {
    pub kanji: String,
    pub original: Option<String>,
    pub phon: Option<String>,
    pub position: Option<Position>,
    pub partial: bool,
    pub number: Option<u8>,
    pub part: Option<u8>,
    pub radical_form: bool,
    pub radical: Option<Radical>,
    pub strokes: Vec<Stroke>,
    pub parts: Vec<Kanji>,
    pub trad_form: bool,
    pub variant: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum Position {
    Bottom,
    Kamae,
    Left,
    Nyo,
    Nyoc,
    Right,
    Tare,
    Tarec,
    Top,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum Radical {
    General,
    Jis,
    Nelson,
    Tradit,
}

impl FromStr for Position {
    type Err = ();

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "bottom" => Ok(Position::Bottom),
            "kamae" => Ok(Position::Kamae),
            "left" => Ok(Position::Left),
            "nyo" => Ok(Position::Nyo),
            "nyoc" => Ok(Position::Nyoc),
            "right" => Ok(Position::Right),
            "tare" => Ok(Position::Tare),
            "tarec" => Ok(Position::Tarec),
            "top" => Ok(Position::Top),
            _ => Err(()),
        }
    }
}

impl FromStr for Radical {
    type Err = ();

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "general" => Ok(Radical::General),
            "jis" => Ok(Radical::Jis),
            "nelson" => Ok(Radical::Nelson),
            "tradit" => Ok(Radical::Tradit),
            _ => Err(()),
        }
    }
}
